bitsandbytes/backends/__init__.py

```diff
@@ -0,0 +1,13 @@
+from typing import Dict
+
+from bitsandbytes.backends.base import Backend
+
+backends: Dict[str, Backend] = {}
+
+def register_backend(backend_name: str, backend_instance: Backend):
+    backends[backend_name.lower()] = backend_instance
+
+def ensure_backend_is_available(device_type: str):
```

Comment:
I think we can have a better design here. Currently, this function is called before we call any backend function. This bloats the function calls. I think it would be better to set the backend once and do not keep a dictionary around. A singleton class would be appropriate for this and we initialize this class through the `cuda_setup` or equivalent `device_setup`. 

This also prevents a different problem: Currently, the dictionary is indexed via the device type (e.g. `A.device.type`), but I am not sure if this is usable for all devices. I am not sure what AMD devices and Apple devices say for this. 

Another problem with this is that currently, paged optimizer buffers have two device types CPU and CUDA where on the PyTorch level only CPU is visible. As such, this would currently be executed on the wrong device. Although paged optimizers have CPU buffers they need to be executed on CUDA device code.

Reply:
I also disagree somewhat with an `ensure_backend_is_available(...)` function – if we want a function that raises a neat exception,

```
def get_backend(device: str) -> Backend:
    backend = backends.get(device)
    if not backend:
        raise UserFriendlyError
    return backend
```
would do, or even `get_backend_for(tensor: ...)`...

If we set _a_ backend once, we can't have a library that would support a system with, say, both a `"cpu"` and `"mps"` backend at the same time.

As for indexing the dictionary, I don't see that being a very pressing issue for just now. The Apple Silicon device type is `"mps"`, FWIW.

Reply:
AMD (HIP) and any other GPU device will say `cuda` as well afaik. But I agree if there isn't a simple solution right now this isn't that urgent. Without cublasLt/hipblasLt support the API's are the same. The only way I know to differentiate between cuda and hip is `torch.version.hip` which you can't get on a per tensor basis. The only idea I have to make a separate hip device possible is instead of e.g.

```python
return backends[A.device.type].dequantize_4bit(...)
```

we add a function

```python
def get_tensor_backend(A: torch.Tensor):
    if A.device.type=="cuda" and torch.version.hip:
        return backends["hip"]
    return backends[A.device.type]
```

and then call 
```python
return get_tensor_backend(A).dequantize_4bit(...)
```

but this would add extra overhead and it might not even be worth it to use a separate backend for the couple functions that need it and instead we just add a `torch.version.hip` check there. We could also include the `ensure_backend_is_available` error handling in there.

Reply:
From what I understand we don't need to support multiple GPU backends at the same time, but at most CPU + GPU. I wonder what would be the best way to handle those two and initialize them only once at library initialization.

Reply:
Why is that ? Isn't multiple GPUs a valid setup?

Reply:
AMD+CUDA seems like a stretch, considering PyTorch uses the same CUDA semantics for ROCm.

As far as Apple goes, one would hope support for whatever the underlying hardware is can come through Metal if future GPU options become available. In fact, as it is [MTLGPUFamily.metal3](https://developer.apple.com/documentation/metal/mtlgpufamily/metal3) already lists out the 2019 Mac Pro (x86-64) options from AMD.

My perspective is that, at least for right now, Intel GPU + (NVIDIA or AMD) is the most viable mixed-vendor GPU setup.

Reply:
Probably. I was more making the argument that in PyTorch you can spin up any number of heterogeneous devices and shuffle tensors between them easily.
Unless there is a very compelling reason not to, I think it would be good to not making assumptions about a single device or a single device type, and the best method for choosing the backend would be to look at what is sent in and load the backend lazily based off of this information.